### DISCUSSION TITLE
How does this compare with PydanticAI?

### DISCUSSION THREAD

**Discussion by zaphod72** [BrainBlend-AI/atomic-agents#45] [General]
Hi,
Pydantic just released [PydanticAI](https://ai.pydantic.dev/).

> a Python Agent Framework designed to make it less painful to build production grade applications with Generative AI.

How does this compare with atomic-agents?
Will atomic-agents look to build on top of PydanticAI or are these two alternate solutions for building agentic AI workflows?

**Comment by Eigenwise** [comment]
Great question!

Since I had a prospective client ask the same question, I have looked into it today...

My conclusion so far is that, despite calling itself a framework, it feels more like a library. Aside from how tool calling works, it feels like it's much closer to being an alternative/replacement for Instructor, rather than Atomic Agents, though.

Now, Atomic Agents uses Instructor internally, and PydanticAI serves as a possible replacement for this, I think.
Today, I think PydanticAI is still very early (validation was a bit wonky, for example, especially in streaming mode, not that it's perfect with Instructor either, but it's better. Also Pydantic doesn't have an out-of-the-box retry mechanism internally like Instructor does, though I'm sure that is planned).

I feel, though, as PydanticAI progresses more and more, it might be worthwhile to consider replacing Instructor with PydanticAI as the "internal engine" for Atomic Agents.

After all, Atomic Agents is less about providing you functionality, and more about providing you with an organizational layer around Instructor's functionality. For example, how everything works with input&output schemas, how docstrings are enforced through usage of the BaseIOSchema, ...

In the end, the impact on Atomic Agents will depend on whether or not Atomic Agents ever goes to other languages. Today, Instructor is available in JS, rust, ruby and python (maybe I missed some)

Meaning, the logic behind Atomic Agents could be ported to any of those languages rather easily (given that I actually get the time or the help to get around to it) - This might become less easy if we adopt PydanticAI internally.

I guess that'll mostly depend on how much extra benefit PydanticAI will bring over Instructor and whether it's enough to make me want to stick with that... In the end, if I can cut out a dependency, like Instructor, without losing functionality since Atomic Agents heavily uses Pydantic, that's be amazing!

**Reply by zaphod72** [reply]
Thank you for the quick reply.
It sound like you have a good plan there. It'll be interesting to see where this flexible project goes.

**Comment by chobitsX** [comment]
I prefer Atomic more，pydanticai just like autogen(want's more but more complex)，Atomic is more free and light

**Reply by Eigenwise** [reply]
Thank you! It really means a lot to me to get that kind of feedback!

Happy holidays!
